src/metabase/models/pulse.clj

```diff
@@ -0,0 +1,159 @@
+(ns metabase.models.pulse
+  (:require [clojure.tools.logging :as log]
+            [korma.core :as k]
+            [korma.db :as kdb]
+            [metabase.db :as db]
+            (metabase.models [card :refer [Card]]
+                             [common :refer [perms-readwrite]]
+                             [hydrate :refer :all]
+                             [interface :refer :all]
+                             [pulse-card :refer [PulseCard]]
+                             [pulse-channel :refer [PulseChannel] :as pulse-channel]
+                             [user :refer [User]])
+            [metabase.util :as u]))
+
+
+(defrecord PulseInstance []
+  ;; preserve normal IFn behavior so things like ((sel :one Database) :id) work correctly
+  clojure.lang.IFn
+  (invoke [this k]
+    (get this k)))
+
+(extend-ICanReadWrite PulseInstance :read :public-perms, :write :public-perms)
+
+
+(defentity Pulse
+  [(k/table :pulse)
+   (hydration-keys pulse)
+   timestamped]
+
+  (pre-insert [_ pulse]
+    (let [defaults {:public_perms perms-readwrite}]
+      (merge defaults pulse)))
+
+  (post-select [_ {:keys [id creator_id] :as pulse}]
+    (map->PulseInstance
+      (u/assoc* pulse
+                :cards    (delay (k/select Card
+                                   (k/join PulseCard (= :pulse_card.card_id :id))
+                                   (k/fields :id :name :description :display)
+                                   (k/where {:pulse_card.pulse_id id})
+                                   (k/order :pulse_card.position :asc)))
+                :channels (delay (db/sel :many PulseChannel (k/where {:pulse_id id})))
+                :creator  (delay (when creator_id (db/sel :one User :id creator_id))))))
+
+  (pre-cascade-delete [_ {:keys [id]}]
+    (db/cascade-delete PulseCard :pulse_id id)
+    (db/cascade-delete PulseChannel :pulse_id id)))
+
+(extend-ICanReadWrite PulseEntity :read :public-perms, :write :public-perms)
+
+
+;; ## Helper Functions
+
+(defn- update-pulse-cards
+  "Update the `PulseCards` for a given PULSE.
+   CARD-IDS should be a definitive collection of *all* IDs of cards for the pulse in the desired order.
+
+   *  If an ID in CARD-IDS has no corresponding existing `PulseCard` object, one will be created.
+   *  If an existing `PulseCard` has no corresponding ID in CARD-IDs, it will be deleted.
+   *  All cards will be updated with a `position` according to their place in the collection of CARD-IDS"
+  {:arglists '([pulse card-ids])}
+  [{:keys [id]} card-ids]
+  {:pre [(integer? id)
+         (coll? card-ids)
+         (every? integer? card-ids)]}
+  ;; first off, just delete any cards associated with this pulse (we add them again below)
+  (k/delete PulseCard (k/where {:pulse_id id}))
```

Comment:
probably better to do
```clojure
(cascade-delete PulseCard, :pulse_id id)
```
here. That triggers the multimethod `pre-cascade-delete` so we can delete dependent objects beforehand if needed.